src/persist/src/operators/persistent_upsert.rs

```diff
@@ -31,24 +40,77 @@ pub trait PersistentUpsert<G, K, V, T> {
     /// a "sequence" of upserts would mean for partially ordered timestamps.
-    fn persistent_upsert(&self, as_of_frontier: Antichain<u64>) -> Stream<G, ((K, V), u64, isize)>
+    ///
+    /// **Note:** This does only persist upserts but not seal them. Use together with `seal()` to
+    /// also seal the persistent collection.
+    fn persistent_upsert(
+        &self,
+        as_of_frontier: Antichain<u64>,
+        persist_config: PersistentUpsertConfig<K, V>,
+    ) -> Stream<G, ((K, V), u64, isize)>
     where
         G: Scope<Timestamp = u64>;
 }
 
+/// Persist configuration for persistent upsert.
+pub struct PersistentUpsertConfig<K: Codec, V: Codec> {
+    /// The timestamp up to which which data should be read when restoring.
+    upper_seal_ts: u64,
+
+    /// [`StreamReadHandle`] for the collection that we should persist to.
+    read_handle: StreamReadHandle<K, V>,
+
+    /// [`StreamWriteHandle`] for the collection that we should persist to.
+    write_handle: StreamWriteHandle<K, V>,
+}
+
+impl<K: Codec, V: Codec> PersistentUpsertConfig<K, V> {
+    /// Creates a new [`PersistentUpsertConfig`] from the given parts.
+    pub fn new(
+        upper_seal_ts: u64,
+        read_handle: StreamReadHandle<K, V>,
+        write_handle: StreamWriteHandle<K, V>,
+    ) -> Self {
+        PersistentUpsertConfig {
+            upper_seal_ts,
+            read_handle,
+            write_handle,
+        }
+    }
+}
+
 impl<G, K, V, T> PersistentUpsert<G, K, V, T> for Stream<G, (K, Option<V>, T)>
 where
     G: Scope<Timestamp = u64>,
-    K: timely::Data + timely::ExchangeData + Hash + Eq,
-    V: timely::Data + timely::ExchangeData,
+    K: timely::Data + timely::ExchangeData + Codec + Debug + Hash + Eq,
+    V: timely::Data + timely::ExchangeData + Codec + Debug + Hash + Eq,
     T: timely::Data + timely::ExchangeData + Ord,
 {
-    fn persistent_upsert(&self, as_of_frontier: Antichain<u64>) -> Stream<G, ((K, V), u64, isize)>
+    fn persistent_upsert(
+        &self,
+        as_of_frontier: Antichain<u64>,
+        persist_config: PersistentUpsertConfig<K, V>,
+    ) -> Stream<G, ((K, V), u64, isize)>
     where
         G: Scope<Timestamp = u64>,
     {
-        let result_stream = self.unary_frontier(
+        let (restored_upsert_oks, _state_errs) = {
+            let snapshot = persist_config
+                .read_handle
+                .snapshot()
+                .expect("cannot take snapshot");
+            let (restored_oks, restored_errs) = operators::replay(&mut self.scope(), snapshot);
+            let uppser_seal_ts = persist_config.upper_seal_ts;
+            let restored_oks = restored_oks.filter(move |(_data, ts, _diff)| *ts < uppser_seal_ts);
+            (restored_oks, restored_errs)
+        };
+
+        let mut differential_state_ingester = Some(DifferentialStateIngester::new());
+
+        let new_upsert_oks = self.binary_frontier(
+            &restored_upsert_oks,
```

Comment:
Hmm actually using replay as an input instead of inlining the logic maybe makes this feel more complicated to me, though I'm unsure about that. What do you think about sharing the guts of the replay operator and keeping this as a unary_frontier operator?

Reply:
The one big thing we get from separating the replay out is that we simply need to `Exchange` both the data and the upsert state to make sure that state ends up on the worker it needs to end up on. If we inlined the replay, we would have to replicate the hashing/shuffling logic that timely uses, and we would be depending on that implementation detail for correctness.

Another thing we're getting is that we only need to read the upsert state once. If each operator did the reading, every operator would have to read all data and pick out the items that are relevant for it's "shard" of the key space.

Reply:
Ah that's right, we talked about this before and I forgot. What you have is definitely the way to go, let's just document why so the next time I forget, you don't have to remind me :)